File: api/models/matiere.js
const mongoose = require("mongoose");

const historiqueSchema = new mongoose.Schema({
    action: { type: String, required: true }, // "ajout" ou "consommation"
    quantite: { type: Number, required: true },
    date: { type: Date, default: Date.now }
  });

const matiereSchema = new mongoose.Schema({
  reference: { type: String, required: true, unique: true },
  couleur: { type: String, required: true },
  quantite: { type: Number, required: true, min: 0 },
  dateAjout: { type: Date, default: Date.now }, 
  historique: [historiqueSchema] 
});

const Matiere = mongoose.model("Matiere", matiereSchema);

module.exports = Matiere;

File: frontend_flutter/lib/models/matiere.dart
class Historique {
  final String action;
  final double quantite;
  final DateTime date;

  Historique({required this.action, required this.quantite, required this.date});

  factory Historique.fromJson(Map<String, dynamic> json) {
    return Historique(
      action: json['action'],
      quantite: (json['quantite'] as num).toDouble(),
      date: DateTime.parse(json['date']),
    );
  }
  Map<String, dynamic> toJson() {
    return {
      'action': action,
       'quantite': quantite,
      'date': date.toIso8601String(),
    };
  }
}
class Matiere {
  final String id;
  String reference;
  final String couleur;
  final double quantite;
  final DateTime dateAjout;
  final List<Historique> historique;

  Matiere({
    required this.id,
    required this.reference,
    required this.couleur,
    required this.quantite,
    required this.dateAjout,
    required this.historique,
  });

  // Convertir un JSON en objet Matiere
  factory Matiere.fromJson(Map<String, dynamic> json) {
    return Matiere(
      id: json['_id']?.toString() ?? '',
    reference: json['reference'] ?? '', 
    couleur: json['couleur'] ?? '',
    quantite:(json['quantite'] as num).toDouble() ?? 0,
    dateAjout: DateTime.parse(json['dateAjout']),
      historique: (json['historique'] as List).map((item) => Historique.fromJson(item)).toList(),
    );
  }

  // Convertir un objet Matiere en JSON
  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'reference': reference,
      'couleur': couleur,
      'quantite': quantite,
      'dateAjout': dateAjout.toIso8601String(),
      'historique': historique.map((h) => h.toJson()).toList(),
    };
  }
  bool estStockSuffisant(double besoin) {
    return quantite >= besoin;
  }
}
